modules/ttdBidAdapter.js

```diff
@@ -0,0 +1,602 @@
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.maxduration must be set to the maximum video ad duration in seconds');
+        return false;
+      }
+      if (!mediaTypesVideo.playerSize) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.playerSize should be an array of width, height values')
+        return false;
+      }
+      if (!mediaTypesVideo.api || mediaTypesVideo.api.length === 0) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.api should be an array of supported api frameworks. See the Open RTB v2.5 spec for valid values');
+        return false;
+      }
+      if (!mediaTypesVideo.mimes || mediaTypesVideo.mimes.length === 0) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.mimes should be an array of supported mime types');
+        return false;
+      }
+      if (!mediaTypesVideo.protocols) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.protocols should be an array of supported protocols. See the Open RTB v2.5 spec for valid values')
+        return false;
+      }
+      if (mediaTypesVideo.startdelay && !utils.isInteger(mediaTypesVideo.startdelay)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.startdelay should be an integer');
+        return false;
+      }
+      if (mediaTypesVideo.skip && !(mediaTypesVideo.skip === 0 || mediaTypesVideo.skip === 1)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.skip should be either 0 or 1');
+        return false;
+      }
+      if (mediaTypesVideo.skipmin && !utils.isInteger(mediaTypesVideo.skipmin)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.skipmin should be an integer');
+        return false;
+      }
+      if (mediaTypesVideo.skipafter && !utils.isInteger(mediaTypesVideo.skipafter)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.skipafter should be an integer');
+        return false;
+      }
+      if (mediaTypesVideo.minbitrate && !utils.isInteger(mediaTypesVideo.minbitrate)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.minbitrate should be an integer');
+        return false;
+      }
+      if (mediaTypesVideo.maxbitrate && !utils.isInteger(mediaTypesVideo.maxbitrate)) {
+        utils.logWarn(BIDDER_CODE + ': mediaTypes.video.maxbitrate should be an integer');
+        return false;
+      }
+    }
+
+    const pageCategories = utils.deepAccess(bid, 'params.categories');
+    if (pageCategories) {
+      if (!Array.isArray(pageCategories)) {
+        utils.logWarn(BIDDER_CODE + ': params.categories should be an array of string values');
+        return false;
+      }
+      for (let i = 0; i < pageCategories.length; i++) {
+        if (typeof pageCategories[i] !== 'string') {
+          utils.logWarn(BIDDER_CODE + ': params.categories should be an array of string values');
+          return false;
+        }
+      }
+    }
+
+    const keywords = utils.deepAccess(bid, 'params.keywords');
```

Comment:
Please prefer openrtb object to your own bidder params; or merge them, eg publishers typically set keywords as per https://docs.prebid.org/features/firstPartyData.html 

eg  

pbjs.setConfig({
   ortb2: {
       site: {
           name: "example",
           domain: "page.example.com",
           cat: ["IAB2"],
           sectioncat: ["IAB2-2"],
           pagecat: ["IAB2-2"],
           page: "https://page.example.com/here.html",
           ref: "https://ref.example.com",
           keywords: "power tools, drills",
           search: "drill",


You should read these keywords, not forcing publishers to duplicate to your adapter params. This simplifies integration a great deal.

Reply:
Thank you, I've added removed `keywords` as a param and am now reading it from the first party data site keywords.